x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/QueryRuleset.java

```diff
@@ -0,0 +1,162 @@
+    }
+
+    public QueryRuleset(StreamInput in) throws IOException {
+        this.id = in.readString();
+        this.rules = in.readList(QueryRule::new);
+    }
+
+    private static final ConstructingObjectParser<QueryRuleset, String> PARSER = new ConstructingObjectParser<>(
+        "query_ruleset",
+        false,
+        (params, resourceName) -> {
+            final String id = (String) params[0];
+            // Check that id matches the resource name. We don't want it to be updatable
+            if (id.equals(resourceName) == false) {
+                throw new IllegalArgumentException(
+                    "Query ruleset identifier [" + id + "] does not match the resource name: [" + resourceName + "]"
+                );
+            }
+            @SuppressWarnings("unchecked")
+            final List<QueryRule> rules = (List<QueryRule>) params[1];
+
+            return new QueryRuleset(resourceName, rules);
+        }
+    );
+
+    public static final ParseField ID_FIELD = new ParseField("ruleset_id");
+    public static final ParseField RULES_FIELD = new ParseField("rules");
+
+    static {
+        PARSER.declareString(constructorArg(), ID_FIELD);
+        PARSER.declareObjectArray(constructorArg(), (p, c) -> QueryRule.fromXContent(p), RULES_FIELD);
+    }
+
+    /**
+     * Parses a {@link QueryRuleset} from its {@param xContentType} representation in bytes.
+     *
+     * @param resourceName The name of the resource (must match the {@link QueryRuleset} id).
+     * @param source The bytes that represents the {@link QueryRuleset}.
+     * @param xContentType The format of the representation.
+     *
+     * @return The parsed {@link QueryRuleset}.
+     */
+    public static QueryRuleset fromXContentBytes(String resourceName, BytesReference source, XContentType xContentType) {
+        try (XContentParser parser = XContentHelper.createParser(XContentParserConfiguration.EMPTY, source, xContentType)) {
+            return QueryRuleset.fromXContent(resourceName, parser);
+        } catch (IOException e) {
+            throw new ElasticsearchParseException("Failed to parse: " + source.utf8ToString(), e);
+        }
+    }
+
+    /**
+     * Parses a {@link QueryRuleset} through the provided {@param parser}.
+     *
+     * @param resourceName The name of the resource (must match the {@link QueryRuleset} name).
+     * @param parser The {@link XContentType} parser.
+     *
+     * @return The parsed {@link QueryRuleset}.
+     */
+    public static QueryRuleset fromXContent(String resourceName, XContentParser parser) throws IOException {
+        return PARSER.parse(parser, resourceName);
+    }
+
+    /**
+     * Converts the {@link QueryRuleset} to XContent.
+     *
+     * @return The {@link XContentBuilder} containing the serialized {@link QueryRuleset}.
+     */
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
```

Comment:
It's because `ToXContent` is considered as a fragment by default and not an entire object. We should use `ToXContentObject` instead so that you have full control (nothing is added externally) on the serialisation of the full object.

Reply:
Thank you!